chia/types/spend_bundle.py

```diff
@@ -77,6 +80,18 @@ def not_ephemeral_additions(self):
 
         return result
 
+    def get_memos(self) -> Dict[bytes32, bytes]:
+        memos: Dict[bytes32, bytes] = {}
+        for coin_spend in self.coin_spends:
+            result = Program.from_bytes(bytes(coin_spend.puzzle_reveal)).run(
+                Program.from_bytes(bytes(coin_spend.solution)))
```

Comment:
I think we should only run the CLVM code once. Either when we consider a `SpendBundle` for inclusion in the mempool, or when validating (and applying the effects of) a block.

It's not clear to me that it's reasonable to be able to ask a `SpendBundle` for its memos. Just like I don't think it's reasonable to be able to ask it for its additions and removals ether. The additions, removals and memos are computed (at a possibly high price) when a Spend bundle is inserted into the mempool (where we save off the additions and removal) and when we validate a block, where we also extract additions and removals in order to apply them to the peak block.

I think it would make more sense to treat memos as a property of "additions" and be accessible in places where we have computed them, not necessarily here.

Reply:
in the full node (i.e. not wallets) the spend bundle program / generator program is executed by `get_name_puzzle_conditions()` and returns `NPCResult`, which includes additions, removals and memos.

Reply:
The NPCs are not stored in the full node though, all that is stored is the coins, and the historical generators. 
The use case is a client that wants access to a memo for a certain coin.
They would do get_puzzle_and_solution for a specific coin, and then get it's memos.

Reply:
Also, we never call .get_memos() outside of the tests right now. It is meant to be called by clients who are using the python RPC library. Maybe we should move it out into the client side instead?